src/coroutine/memoset/prove.rs

```diff
@@ -0,0 +1,407 @@
+        for ptr in output_ptrs {
+            output.push(ptr.tag().clone());
+            output.push(ptr.hash().clone());
+        }
+        Ok((next_pc, output))
+    }
+
+    fn circuit_index(&self) -> usize {
+        self.query_index
+    }
+}
+
+impl<'a, F: CurveCycleEquipped, Q: Query<F> + Send + Sync>
+    RecursiveSNARKTrait<F, Coroutine<'a, F, Q>> for Proof<F, Coroutine<'a, F, Q>>
+{
+    type PublicParams = PublicParams<F>;
+
+    type ErrorType = SuperNovaError;
+
+    #[tracing::instrument(skip_all, name = "supernova::prove_recursively")]
+    fn prove_recursively(
+        pp: &PublicParams<F>,
+        z0: &[F],
+        steps: Vec<Coroutine<'a, F, Q>>,
+        _store: &Store<F>,
+    ) -> Result<Self, ProofError> {
+        let mut recursive_snark_option: Option<RecursiveSNARK<E1<F>>> = None;
+
+        let z0_primary = z0;
+        let z0_secondary = Self::z0_secondary();
+
+        let mut prove_step = |i: usize, step: &Coroutine<'a, F, Q>| {
+            info!("prove_recursively, step {i}");
+
+            let secondary_circuit = step.secondary_circuit();
+
+            let mut recursive_snark = recursive_snark_option.clone().unwrap_or_else(|| {
+                info!("prove_recursively, step {i}");
+                RecursiveSNARK::new(
+                    &pp.pp,
+                    step,
+                    step,
+                    &secondary_circuit,
+                    z0_primary,
+                    &z0_secondary,
+                )
+                .unwrap()
+            });
+
+            recursive_snark
+                .prove_step(&pp.pp, step, &secondary_circuit)
+                .unwrap();
+
+            recursive_snark_option = Some(recursive_snark);
+        };
+        for (i, step) in steps.iter().enumerate() {
+            prove_step(i, step);
+        }
+        // This probably should be made unnecessary.
+        Ok(Self::Recursive(
+            Box::new(recursive_snark_option.expect("RecursiveSNARK missing")),
+            PhantomData,
+        ))
+    }
+
+    fn compress(self, pp: &PublicParams<F>) -> Result<Self, ProofError> {
+        match &self {
+            Self::Recursive(recursive_snark, _) => {
+                let snark = CompressedSNARK::prove(&pp.pp, pp.pk(), recursive_snark)?;
+                Ok(Self::Compressed(Box::new(snark), PhantomData))
+            }
+            Self::Compressed(..) => Ok(self),
+        }
+    }
+
+    fn verify(&self, pp: &Self::PublicParams, z0: &[F], zi: &[F]) -> Result<bool, Self::ErrorType> {
+        let (z0_primary, zi_primary) = (z0, zi);
+        let z0_secondary = Self::z0_secondary();
+        let zi_secondary = &z0_secondary;
+
+        let (zi_primary_verified, zi_secondary_verified) = match self {
+            Self::Recursive(p, _) => p.verify(&pp.pp, z0_primary, &z0_secondary)?,
+            Self::Compressed(p, _) => p.verify(&pp.pp, pp.vk(), z0_primary, &z0_secondary)?,
+        };
```

Comment:
It seems like some of this could be moved to default implementations or something. I can believe it's necessary to include here, but it seems like duplication we should be able to avoid somehow.

Reply:
That would require a trait to unify these two functions, we can definitely come up with it afterwards